Marlin/planner.cpp

```diff
@@ -600,45 +627,56 @@ void Planner::calculate_volumetric_multipliers() {
 
   void Planner::unapply_leveling(float raw[XYZ]) {
 
-    if (!leveling_active) return;
+    #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
+      const float fade_scaling_factor = fade_scaling_factor_for_z(raw[Z_AXIS]);
+    #else
+      constexpr float fade_scaling_factor = 1.0;
+    #endif
 
-    #if ABL_PLANAR
+    if (leveling_active && fade_scaling_factor) {
 
-      matrix_3x3 inverse = matrix_3x3::transpose(bed_level_matrix);
+      #if ABL_PLANAR
 
-      float dx = raw[X_AXIS] - (X_TILT_FULCRUM),
-            dy = raw[Y_AXIS] - (Y_TILT_FULCRUM);
+        matrix_3x3 inverse = matrix_3x3::transpose(bed_level_matrix);
 
-      apply_rotation_xyz(inverse, dx, dy, raw[Z_AXIS]);
+        float dx = raw[X_AXIS] - (X_TILT_FULCRUM),
+              dy = raw[Y_AXIS] - (Y_TILT_FULCRUM);
 
-      raw[X_AXIS] = dx + X_TILT_FULCRUM;
-      raw[Y_AXIS] = dy + Y_TILT_FULCRUM;
+        apply_rotation_xyz(inverse, dx, dy, raw[Z_AXIS]);
 
-    #else
+        raw[X_AXIS] = dx + X_TILT_FULCRUM;
+        raw[Y_AXIS] = dy + Y_TILT_FULCRUM;
 
-      #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
-        const float fade_scaling_factor = fade_scaling_factor_for_z(raw[Z_AXIS]);
-        if (!fade_scaling_factor) return;
-      #elif HAS_MESH
-        constexpr float fade_scaling_factor = 1.0;
-      #endif
+      #else // !ABL_PLANAR
 
-      raw[Z_AXIS] -= (
-        #if ENABLED(AUTO_BED_LEVELING_UBL)
-          ubl.get_z_correction(raw[X_AXIS], raw[Y_AXIS]) * fade_scaling_factor
-        #elif ENABLED(MESH_BED_LEVELING)
-          mbl.get_z(raw[X_AXIS], raw[Y_AXIS]
-            #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
-              , fade_scaling_factor
-            #endif
-          )
-        #elif ENABLED(AUTO_BED_LEVELING_BILINEAR)
-          bilinear_z_offset(raw) * fade_scaling_factor
-        #else
-          0
-        #endif
-      );
+        raw[Z_AXIS] -= (
+          #if ENABLED(AUTO_BED_LEVELING_UBL)
+            ubl.get_z_correction(raw[X_AXIS], raw[Y_AXIS]) * fade_scaling_factor
+          #elif ENABLED(MESH_BED_LEVELING)
+            mbl.get_z(raw[X_AXIS], raw[Y_AXIS]
+              #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
+                , fade_scaling_factor
+              #endif
+            )
+          #elif ENABLED(AUTO_BED_LEVELING_BILINEAR)
+            bilinear_z_offset(raw) * fade_scaling_factor
+          #else
+            0
+          #endif
+        );
 
+      #endif // !ABL_PLANAR
+    }
+
+    #if ENABLED(SKEW_CORRECTION)
+      if (WITHIN(raw[X_AXIS], X_MIN_POS, X_MAX_POS) && WITHIN(raw[Y_AXIS], Y_MIN_POS, Y_MAX_POS)) {
+        const float temprx = raw[X_AXIS] + raw[Y_AXIS] * planner.xy_skew_factor + raw[Z_AXIS] * planner.xz_skew_factor,
```

Comment:
@Paciente8159 — I wanted to double-check that this line is correct. Unlike the line in `apply_leveling` above, it doesn't include `yz_skew_factor`. Should it instead be…
```cpp
const float temprx = raw[X_AXIS] + raw[Y_AXIS] * planner.xy_skew_factor + raw[Z_AXIS] * (planner.xz_skew_factor - (planner.xy_skew_factor * planner.yz_skew_factor));
```
Thanks for clearing this up. Probably no one has noticed this yet, if it is an issue.

Reply:
Sorry for the late reply.
I checked my local repo code.
It's almost that (- sign and not a + sign when factoring XY compensation). It should be like this:

`const float temprx = raw[X_AXIS] - (raw[Y_AXIS] * planner.xy_skew_factor) - (raw[Z_AXIS] * (planner.xz_skew_factor - (planner.xy_skew_factor * planner.yz_skew_factor)));`

Reply:
@thinkyhead 
I've checked the config file and after doing some math the signs must be negative.
For example if the internal angle (lower left at corner A) is smaller than 90º the skew factor will be positive. X will have to be decremented to compensate for the angle.

Reply:
The code I'm pointing out is `unskew` — reversal of the `skew`.

What I currently have in the code…
```cpp
#if ENABLED(SKEW_CORRECTION)

  FORCE_INLINE static void skew(float &cx, float &cy, const float &cz) {
    if (WITHIN(cx, X_MIN_POS + 1, X_MAX_POS) && WITHIN(cy, Y_MIN_POS + 1, Y_MAX_POS)) {
      const float sx = cx - (cy * xy_skew_factor) - (cz * (xz_skew_factor - (xy_skew_factor * yz_skew_factor))),
                  sy = cy - (cz * yz_skew_factor);
      if (WITHIN(sx, X_MIN_POS, X_MAX_POS) && WITHIN(sy, Y_MIN_POS, Y_MAX_POS)) {
        cx = sx; cy = sy;
      }
    }
  }

  FORCE_INLINE static void unskew(float &cx, float &cy, const float &cz) {
    if (WITHIN(cx, X_MIN_POS, X_MAX_POS) && WITHIN(cy, Y_MIN_POS, Y_MAX_POS)) {
      const float ux = cx + cy * xy_skew_factor + cz * xz_skew_factor,
                  uy = cy + cz * yz_skew_factor;
      if (WITHIN(ux, X_MIN_POS, X_MAX_POS) && WITHIN(uy, Y_MIN_POS, Y_MAX_POS)) {
        cx = ux; cy = uy;
      }
    }
  }

#endif // SKEW_CORRECTION
```
The line that does the "unskew" for X uses your original code:
```cpp
ux = cx + cy * XY + cz * XZ;
```
But I think maybe it should be this:
```cpp
ux = cx + cy * XY + cz * (XZ + (XY * XZ));
```
Are you saying it should actually be this?
```cpp
ux = cx + cy * XY + cz * (XZ - (XY * XZ));
```

Reply:
Ahh OK.
The correct way is:
```
ux = cx + cy * XY + cz * XZ;
uy = cy + cz * YZ;
```
And the reverse operation of this is:
```
sx = cx - (cy * XY) - (cz * (XZ- (XY* YZ)));
sy = cy - (cz * YZ);
```

This is the result of applying matrix operations:
The unskew matrix - the matrix with the factors computed from the real machine distortion of what should be a perfect cube;
And the skew matrix - the math equation equivalent to the inverted unskew matrix.

You can see this in action in OpenSCAD by comenting the top multmatrix on this code snippet:
```
//Some random values for the skew factor
XY=0.2;
XZ=0.1;
YZ=-0.3;

multmatrix(m = [ [1, -XY, -(XZ-(XY*YZ)), 0],
                 [0, 1, -YZ, 0],
                 [0, 0, 1, 0],
                 [0, 0, 0,  1]
              ])
multmatrix(m = [ [1, XY, XZ, 0],
                 [0, 1, YZ, 0],
                 [0, 0, 1, 0],
                 [0, 0, 0,  1]
              ])
cube(size=[100,100,100],center=false);
```
This is why I say that bed leveling can affect skew and vice versa. The math is not that direct and changes on one axis affect all axis. Matrix operations are a more effective/correct way to perform 3D machine to world transformations although for small distortions this can be acceptable.

Reply:
Makes sense, thanks!